File: src/resolver.ts
/* Copyright (c) 2024-2025 Voxgig, MIT License */


import Path from 'node:path'


async function resolveElements(
  ctx: any,
  kind: string,
  subkind: string,
  standard: Record<string, any>
) {

  const { log, model, guide } = ctx

  // TODO: model access should be via a utility that generates
  // useful errors when the target is missing
  const control = guide.control[kind][subkind]

  const target = kind + '.' + subkind

  const elementNames = control.order
    .split(/\s*,\s*/)
    .map((t: string) => t.trim())
    .filter((elem: string) => '' != elem)

  log.info({
    point: 'element', note: target + ': order: ' + elementNames.join(';'),
    order: elementNames
  })

  const elementResults: any = []

  for (const en of elementNames) {
    log.debug({ point: 'element', note: target + ': ' + en })
    const element = await resolveElement(en, control, target, standard, ctx)
    const result = await element(ctx)
    elementResults.push(result)
  }

  return elementResults
}


async function resolveElement(
  en: string,
  control: any,
  target: string,
  standard: Record<string, any>,
  ctx: any
) {
  const { log } = ctx

  let element = standard[en]
  if (element) {
    return element
  }

  const elemdef = control.element[en]
  if (null == elemdef) {
    const err = new Error('Unknown element: ' + en)
    log.error({ what: 'element', element: target + ': ' + en, fail: 'unknown', err })
    throw err
  }

  if (!en.startsWith('custom')) {
    const err =
      new Error('Custom element name must start with "custom": ' + en)
    log.error({ what: 'element', element: target + ': ' + en, fail: 'prefix', err })
    throw err
  }

  const customtpath = Path.join(ctx.defpath, elemdef.load)
  try {
    const elementModule = require(customtpath)
    element = elementModule[en]
  }
  catch (e: any) {
    const err = new Error('Custom element not found: ' +
      customtpath + ': ' + e.message)
    log.error({ what: 'element', element: target + ': ' + en, fail: 'require', err })
    throw err
  }

  return element
}


export {
  resolveElements,
}

File: go/resolver.go
/* Copyright (c) 2024-2025 Voxgig, MIT License */

package apidef

import (
	"fmt"
	"path/filepath"
)

// ResolveElements resolves and executes guide elements in order.
func ResolveElements(
	ctx *ApiDefContext,
	kind string,
	subkind string,
	standard map[string]func(*ApiDefContext) (any, error),
) ([]any, error) {
	guide := ctx.Guide
	controlMap, _ := guide["control"].(map[string]any)
	kindMap, _ := controlMap[kind].(map[string]any)
	subkindMap, _ := kindMap[subkind].(map[string]any)

	target := kind + "." + subkind

	orderStr, _ := subkindMap["order"].(string)
	elementNames := splitAndFilterComma(orderStr)

	results := make([]any, 0)

	for _, en := range elementNames {
		element, err := resolveElement(en, subkindMap, target, standard, ctx)
		if err != nil {
			return nil, err
		}
		result, err := element(ctx)
		if err != nil {
			return nil, err
		}
		results = append(results, result)
	}

	return results, nil
}

func resolveElement(
	en string,
	control map[string]any,
	target string,
	standard map[string]func(*ApiDefContext) (any, error),
	ctx *ApiDefContext,
) (func(*ApiDefContext) (any, error), error) {
	// Check standard elements first
	if elem, ok := standard[en]; ok {
		return elem, nil
	}

	// Check control definition
	elementMap, _ := control["element"].(map[string]any)
	elemdef, _ := elementMap[en].(map[string]any)
	if elemdef == nil {
		return nil, fmt.Errorf("unknown element: %s", en)
	}

	// Custom elements must start with "custom"
	if len(en) < 6 || en[:6] != "custom" {
		return nil, fmt.Errorf("custom element name must start with \"custom\": %s", en)
	}

	// Load custom element (Go plugin loading not supported - return error)
	loadPath, _ := elemdef["load"].(string)
	customPath := filepath.Join(ctx.DefPath, loadPath)
	return nil, fmt.Errorf("custom element loading not supported in Go: %s (%s)",
		en, RelativizePath(customPath))
}

func splitAndFilterComma(s string) []string {
	var result []string
	for _, part := range splitByComma(s) {
		trimmed := trimSpace(part)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

func splitByComma(s string) []string {
	var result []string
	start := 0
	for i := range s {
		if s[i] == ',' {
			result = append(result, s[start:i])
			start = i + 1
		}
	}
	result = append(result, s[start:])
	return result
}

func trimSpace(s string) string {
	start, end := 0, len(s)
	for start < end && (s[start] == ' ' || s[start] == '\t' || s[start] == '\n' || s[start] == '\r') {
		start++
	}
	for end > start && (s[end-1] == ' ' || s[end-1] == '\t' || s[end-1] == '\n' || s[end-1] == '\r') {
		end--
	}
	return s[start:end]
}
